### DISCUSSION TITLE
All TCP messages get filtered in `Ip6::PassToHost`

### DISCUSSION THREAD

**Discussion by Irving-cl** [openthread/openthread#10878] [General]
With current implementation of ip6 filtering and TCP handling, when `OPENTHREAD_CONFIG_TCP_ENABLE`  is enabled, TCP messages will only be handled by the OT stack and will never be passed to the host.

In `Ip6::PassToHost`:
```
#if OPENTHREAD_CONFIG_TCP_ENABLE
        // Do not pass TCP message to avoid dual processing from both
        // OpenThread and POSIX TCP stacks.
        case kProtoTcp:
            error = kErrorNoRoute;
            ExitNow();
#endif
```

There are some cases where we hope the TCP messages being passed to the host. For example, I'm using `iperf3`. The iperf client will first send a TCP message when starting the test. With current implementation, the message will be blocked here and not go to the host. Shall we consider to add some mechanism to allow part of the TCP messages go to the host while the others will be handled by OT stack.

@abtink @samkumar

**Comment by jwhui** [comment]
So far, we have not had a use case for needing to handle TCP sessions inside of OT. That's why we simply disable TCP handling within OT on posix-based systems.

If we have specific use cases for needing to handle TCP sessions both inside and outside of OT, then we'll need to figure out how to implement that.

**Reply by Irving-cl** [reply]
Make sense. I'm actually working on NCP and considering if we should disable TCP in NCP build. 

When I disabled `OT_TCP` in NCP, I found that `OPENTHREAD_CONFIG_DNS_CLIENT_OVER_TCP_ENABLE` depends on it. But as I understand, DNS client is mainly for single-chip cases (like cli). When we use NCP, there must be a host. And a seperate dns client should run on the host. So I think we should disable both of them on NCP. Thoughts?

**Reply by jwhui** [reply]
Yes, I think we can rely on DNS client on host for now.
